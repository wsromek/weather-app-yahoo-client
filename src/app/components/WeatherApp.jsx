/* eslint no-console: ['error', { allow: ['error'] }] */

import React from 'react';
import PropTypes from 'prop-types';

import ForecastDisplay from './Forecast/ForecastDisplay';
import CitySelector from './CitySelector';
import LoadingMask from './Util/LoadingMask';
import ErrorMessage from './Util/ErrorMessage';

class WeatherApp extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      error: {
        message: '',
      },
      loading: false,
      cities: props.cities,
      selectedCity: {},
      forecast: {},
    };

    this.apiClient = props.apiClient;
  }

  onForecastFetch = () => {
    const selectedCity = this.state.selectedCity;
    if (!selectedCity.zipCode || !selectedCity.name) {
      return this.setState({
        error: {
          message: 'Please select city first!',
        },
      });
    }

    this.setState({
      loading: true,
    });

    return this.apiClient
      .getWeatherForCity(selectedCity)
      .then((data) => {
        this.setState({
          forecast: data[0],
          error: {
            message: '',
          },
          loading: false,
        });
      })
      .catch((error) => {
        console.error(error);

        this.setState({
          error,
          loading: false,
        });
      });
  };

  onCityChange = (event) => {
    this.setState({
      selectedCity: {
        zipCode: event.target.value.split(':')[0],
        name: event.target.value.split(':')[1],
      },
    });
  };

  onErrorReset = () => {
    this.setState({
      error: false,
    });
  };

  render() {
    return (
      <div>
        <LoadingMask
          loading={this.state.loading}
        />
        <ErrorMessage
          error={this.state.error}
          onErrorReset={this.onErrorReset}
        />
        <CitySelector
          cities={this.state.cities}
          selectedCity={this.state.selectedCity}
          onForecastFetch={this.onForecastFetch}
          onCityChange={this.onCityChange}
        />
        <ForecastDisplay
          forecast={this.state.forecast}
        />
      </div>
    );
  }
}

WeatherApp.propTypes = {
  cities: PropTypes.arrayOf(PropTypes.shape({
    zipCode: PropTypes.string.string,
    name: PropTypes.string.isRequired,
  })).isRequired,
  apiClient: PropTypes.shape({
    getWeatherForCity: PropTypes.func.isRequired,
  }).isRequired,
};

export default WeatherApp;
