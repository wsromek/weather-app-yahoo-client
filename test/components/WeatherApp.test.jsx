import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import forecastFixture from '../fixtures/forecast';
import WeatherApp from '../../src/app/components/WeatherApp';

const cities = [
  {
    zipCode: '35801',
    name: 'Huntsville',
  },
  {
    zipCode: '99501',
    name: 'Anchorage',
  },
];

const apiClientStub = {
  getWeatherForCity: jest.fn().mockReturnValue(
    new Promise(resolve => process.nextTick(() => resolve([forecastFixture]))),
  ),
};

describe('WeatherApp component', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <WeatherApp cities={cities} apiClient={apiClientStub} />,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('fetches weather for selected city when "Get forecast" is clicked', () => {
    const component = mount(<WeatherApp cities={cities} apiClient={apiClientStub} />);

    component.find('select').simulate('change', {
      target: {
        value: '99501:Anchorage',
      },
    });

    component.find('button').simulate('click');

    expect(apiClientStub.getWeatherForCity).toHaveBeenCalledWith({
      zipCode: '99501',
      name: 'Anchorage',
    });
  });
});
