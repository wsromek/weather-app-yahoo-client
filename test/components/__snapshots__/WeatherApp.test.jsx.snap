// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherApp component renders correctly 1`] = `
<div>
  <section
    className="selection"
  >
    <nav
      className="navbar navbar-default"
    >
      <div
        className="container"
      >
        <div
          className="navbar-header"
        >
          <span
            className="navbar-brand"
          >
            Weather Forecast
          </span>
        </div>
        <div
          className="navbar-form navbar-right"
        >
          <div
            className="form-group"
          >
            <select
              className="form-control"
              id="city-list"
              name="city-list"
              onChange={[Function]}
              placeholder="Choose city"
              type={undefined}
            >
              <option
                value="0"
              >
                Choose city:
              </option>
              <option
                value="35801:Huntsville"
              >
                Huntsville
              </option>
              <option
                value="99501:Anchorage"
              >
                Anchorage
              </option>
            </select>
          </div>
          <button
            className="btn btn-primary"
            disabled={false}
            onClick={[Function]}
            style={
              Object {
                "marginLeft": "5px",
              }
            }
            type="button"
          >
            Get forecast
          </button>
        </div>
      </div>
    </nav>
  </section>
</div>
`;
